File: src/model/Person.java
package model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import model.item.ItemType;

import component.PartyComponentDataSource;

import core.Logger;

/**
 * Person consists of a list of skills the person is proficient with as well
 * as their name, and what their profession is called.
 */
@SuppressWarnings("serial")
public class Person implements Conditioned, Inventoried, PartyComponentDataSource, Serializable {
	private final Condition skillPoints;
	
	private final Condition health;

	private boolean dead;
	
	private boolean isMale = true;
	
	private final List<Skill> skills = new ArrayList<Skill>();
	
	private String name;
	
	private Profession profession;
		
	private static final int BASE_SKILL_POINTS = 70;
	
	private final Inventory inventory;

	private double weight;
	
	public static final int MAX_INVENTORY_SIZE = 5;
	
	private static final double MAX_INVENTORY_WEIGHT = 100;
	
	/**
	 * Person creation should be done this way always - 
	 * name first, then profession and skills are added.
	 * @param name 
	 */
	public Person(String name){
		this.name = name;
		this.skillPoints = new Condition(0, BASE_SKILL_POINTS, 0);
		this.health = new Condition(100);
		Logger.log(name + " was created", Logger.Level.DEBUG);
		this.inventory = new Inventory(MAX_INVENTORY_SIZE, MAX_INVENTORY_WEIGHT);
		Random random = new Random();
		this.weight = random.nextInt(100) + 90;
	}
	
	/**
	 * Establish this party member's profession
	 * @param profession the desired profession
	 */
	public void setProfession(Profession profession) {
		//This will only happen during party creation when the skills haven't been chosen yet.
		if (this.profession == null) {
			this.profession = profession;
			Logger.log(this.name + " became a " + this.profession, Logger.Level.DEBUG);
			addSkill(profession.getStartingSkill());
			inventory.clear();
			if(profession.getStartingItem() != null) {
				addItemToInventory(new Item(profession.getStartingItem()));
			}
			return;
		} else if (this.profession == profession) {
			Logger.log(this.name + " is already a " + profession, Logger.Level.DEBUG);
			return;
		} else if (this.profession != profession) {
			this.skills.clear();
			Logger.log(this.name + " stopped being a " + 
					this.profession + " and lost all current skills", Logger.Level.DEBUG);
			this.profession = null;
			setProfession(profession);
			inventory.clear();
			if(profession.getStartingItem() != null) {
				addItemToInventory(new Item(profession.getStartingItem()));
			}
			return;
		} else {
			Logger.log("Don't know what happened", Logger.Level.DEBUG);
			return;
		}
	}

	/**
	 * Gets remaining skill points for consideration during skill adding
	 * @return Remaining skill points
	 */
	public int getSkillPoints() {
		return (int) skillPoints.getCurrent();
	}
	
	/**
	 * Clears the persons skills
	 */
	public void clearSkills() {
		for (int i = skills.size() - 1; i >= 0; i--) {
			Skill skill = skills.get(i);
			if (!skill.equals(profession.getStartingSkill())) {
				skills.remove(skill);
			}
		}
	}
	
	/**
	 * Adds a new skill to the person
	 * @param newSkill The skill to add
	 */
	public void addSkill(Skill newSkill) {
		if(skills.contains(newSkill)) {
			Logger.log(this.name + " already has the skill " 
					+ newSkill, Logger.Level.DEBUG);
			return;
		} else if(newSkill == this.profession.getStartingSkill()) {
			skills.add(newSkill);
			Logger.log("As a " + this.profession + " " + 
					this.name + " gains the skill " + newSkill, Logger.Level.DEBUG);
			return;
		} else {
			skills.add(newSkill);
			Logger.log(this.name + " gained the skill " + newSkill, Logger.Level.DEBUG);
			return;
		}
	}

	/**
	 * Adds a skill to this person's skill set
	 * @param newSkill skill to be added
	 */
	public void buySkill(Skill newSkill){
		if(skills.contains(newSkill)) {
			Logger.log(this.name + " already has the skill " + 
					newSkill, Logger.Level.DEBUG);
			return;
		}
		else {
			if(skillPoints.getCurrent() < newSkill.getCost()) {
				Logger.log(this.name + " does not have enough skill points to obtain " + 
						newSkill + ".  Current skill points: " + 
						skillPoints.getCurrent() + " Cost of new skill: " + 
						newSkill.getCost(), Logger.Level.DEBUG);
			}
			else {
				skillPoints.decrease(newSkill.getCost());
				skills.add(newSkill);
				Logger.log(this.name + " gained the skill " +
						newSkill, Logger.Level.DEBUG);
				Logger.log(this.name + " has " + 
						skillPoints.getCurrent() + " skill points remaning.",
						Logger.Level.DEBUG);
				return;
			}
		}
		return;
	}
	
	/**
	 * Removes skill from person's skill set	
	 * @param oldSkill skill to be removed
	 */
	public void removeSkill(Skill oldSkill) {
		if (!this.skills.contains(oldSkill)) {
			Logger.log("Cannot remove a skill that isn't already known.",
					Logger.Level.DEBUG);
			return;
		}
		else if (this.profession.getStartingSkill() == oldSkill) {
			Logger.log("Cannot remove profession's starting skill", Logger.Level.DEBUG);
			return;
		}
		else {
			this.skills.remove(oldSkill);
			this.skillPoints.increase(oldSkill.getCost());
			return;
		}
	}
	
	/**
	 * Gets the list of skills a person has.
	 * @return The list of skills
	 */
	public List<Skill> getSkills() {
		return skills;
	}
	
	@Override
	public String getName() {
		return name;
	}
	
	/**
	 * Sets the name of the person
	 * @param name The name of the person
	 */
	public void setName(String name) {
		this.name = name;
		Logger.log("Name changed to " + name, Logger.Level.DEBUG);
	}
	
	/**
	 * Profession of person in party
	 * @return Person's profession
	 */
	public Profession getProfession(){
		return profession;
	}

	@Override
	public String toString() {
		final StringBuffer str = new StringBuffer("Name: " + name);
		if(profession != null) {
			 str.append(", Profession: " + profession.getName());
		}
		for(Skill skill : skills) {
			str.append(", Skill: " + skill.getName());
		}
		str.append((isMale ? ", Gender: Male" : ", Gender: Female"));
		return str.toString();
	}
	
	/**
	 * Sets the gender of the person
	 * @param isMale Whether this Person is male or not
	 */
	
	public void setIsMale(boolean isMale) {
		this.isMale = isMale;
		
		Logger.log("Gender changed to " + 
				(isMale ? "Male" : "Female"), Logger.Level.DEBUG);
	}
	
	/**
	 * Returns this person's gender
	 * @return Whether or not this Person is male
	 */
	
	public boolean isMale() {
		return isMale;
	}
	
	@Override
	public Inventory getInventory() {
		return inventory;
	}
	
	@Override
	public void addItemsToInventory(List<Item> items) {
		inventory.addItemsToInventory(items);
	}
	
	@Override
	public void addItemToInventory(Item item) {
		inventory.addItemToInventory(item);
	}

	@Override
	public List<Item> removeItemFromInventory(ItemType itemIndex, int quantity) {
		return inventory.removeItemFromInventory(itemIndex, quantity);
	}
	
	/**
	 * Returns the health condition for status bar.
	 * @return The condition 'health'
	 */
	public Condition getHealth() {
		return health;
	}
	
	/**
	 * Decreases health by amount
	 * @param amount The amount
	 */
	public void decreaseHealth(int amount) {
		health.decrease(amount);
	}
	
	/**
	 * Increases health by amount
	 * @param amount The amount
	 */
	public void increaseHealth(int amount) {
		health.increase(amount);
	}
	
	@Override
	public double getConditionPercentage() {
		return health.getPercentage();
	}
	
	@Override
	public boolean canGetItem(ItemType itemType, int numberOf) {
		return inventory.canGetItems(itemType, numberOf);
	}

	@Override
	public int getMaxSize() {
		return MAX_INVENTORY_SIZE;
	}
	
	@Override
	public double getMaxWeight() {
		return MAX_INVENTORY_WEIGHT;
	}
	
	@Override
	public double getWeight() {
		return inventory.getWeight();
	}

	public void increaseSkillPoints(int amount) {
		skillPoints.increase(amount);
		
	}

	public List<Item> killForFood() {
		int numberOf = (int) (this.weight / ItemType.STRANGEMEAT.getWeight());
		List<Item> itemList = new ArrayList<Item>();
		for(int i = 0; i < (numberOf / 2); i++) {
			itemList.add(new Item(ItemType.STRANGEMEAT));
		}
		return itemList;
	}

	@Override
	public Condition getCondition() {
		return health;
	}

	@Override
	public boolean isDead() {
		return dead;
	}
	
	public void setDead(boolean dead) {
		this.dead = dead;
	}
}

File: web/src/model/Person.ts
import { Condition } from "./Condition";
import { Inventory } from "./Inventory";
import { Item } from "./Item";
import { ItemType } from "./ItemType";
import {
  Profession,
  getProfessionStartingSkill,
  getProfessionStartingItem,
  getProfessionName,
} from "./Profession";
import { Skill, getSkillCost, getSkillName } from "./Skill";

export class Person {
  private readonly skillPoints: Condition;
  private readonly health: Condition;
  private dead: boolean;
  private isMale: boolean;
  private isLeader: boolean;
  private readonly skills: Skill[];
  private name: string;
  private profession: Profession | null;
  private readonly inventory: Inventory;
  private weight: number;

  static readonly BASE_SKILL_POINTS = 70;
  static readonly MAX_INVENTORY_SIZE = 5;
  static readonly MAX_INVENTORY_WEIGHT = 100;

  constructor(name: string) {
    this.name = name;
    this.skillPoints = new Condition(0, Person.BASE_SKILL_POINTS, 0);
    this.health = new Condition(100);
    this.dead = false;
    this.isMale = true;
    this.isLeader = false;
    this.skills = [];
    this.profession = null;
    this.inventory = new Inventory(
      Person.MAX_INVENTORY_SIZE,
      Person.MAX_INVENTORY_WEIGHT,
    );
    this.weight = Math.floor(Math.random() * 100) + 90;
    console.debug(`${name} was created`);
  }

  setProfession(profession: Profession): void {
    if (this.profession === null) {
      this.profession = profession;
      console.debug(`${this.name} became a ${getProfessionName(profession)}`);
      this.addSkill(getProfessionStartingSkill(profession));
      this.inventory.clear();
      const startingItem = getProfessionStartingItem(profession);
      if (startingItem) {
        this.addItemToInventory(new Item(startingItem as unknown as ItemType));
      }
      return;
    }

    if (this.profession === profession) {
      console.debug(
        `${this.name} is already a ${getProfessionName(profession)}`,
      );
      return;
    }

    this.skills.length = 0;
    console.debug(
      `${this.name} stopped being a ${getProfessionName(this.profession)} and lost all current skills`,
    );
    this.profession = null;
    this.setProfession(profession);
    this.inventory.clear();
    const startingItem = getProfessionStartingItem(profession);
    if (startingItem) {
      this.addItemToInventory(new Item(startingItem as unknown as ItemType));
    }
  }

  getSkillPoints(): number {
    return Math.floor(this.skillPoints.getCurrent());
  }

  clearSkills(): void {
    for (let i = this.skills.length - 1; i >= 0; i--) {
      const skill = this.skills[i];
      if (
        this.profession &&
        skill !== getProfessionStartingSkill(this.profession)
      ) {
        this.skills.splice(i, 1);
      }
    }
  }

  addSkill(newSkill: Skill): void {
    if (this.skills.includes(newSkill)) {
      console.debug(
        `${this.name} already has the skill ${getSkillName(newSkill)}`,
      );
      return;
    }

    if (this.profession && newSkill === getProfessionStartingSkill(this.profession)) {
      this.skills.push(newSkill);
      console.debug(
        `As a ${getProfessionName(this.profession)} ${this.name} gains the skill ${getSkillName(newSkill)}`,
      );
      return;
    }

    this.skills.push(newSkill);
    console.debug(`${this.name} gained the skill ${getSkillName(newSkill)}`);
  }

  buySkill(newSkill: Skill): void {
    if (this.skills.includes(newSkill)) {
      console.debug(
        `${this.name} already has the skill ${getSkillName(newSkill)}`,
      );
      return;
    }

    const cost = getSkillCost(newSkill);
    if (this.skillPoints.getCurrent() < cost) {
      console.debug(
        `${this.name} does not have enough skill points to obtain ${getSkillName(newSkill)}. Current skill points: ${this.skillPoints.getCurrent()} Cost of new skill: ${cost}`,
      );
      return;
    }

    this.skillPoints.decrease(cost);
    this.skills.push(newSkill);
    console.debug(`${this.name} gained the skill ${getSkillName(newSkill)}`);
    console.debug(
      `${this.name} has ${this.skillPoints.getCurrent()} skill points remaining.`,
    );
  }

  removeSkill(oldSkill: Skill): void {
    if (!this.skills.includes(oldSkill)) {
      console.debug("Cannot remove a skill that isn't already known.");
      return;
    }

    if (
      this.profession &&
      getProfessionStartingSkill(this.profession) === oldSkill
    ) {
      console.debug("Cannot remove profession's starting skill");
      return;
    }

    const index = this.skills.indexOf(oldSkill);
    if (index > -1) {
      this.skills.splice(index, 1);
    }
    this.skillPoints.increase(getSkillCost(oldSkill));
  }

  getSkills(): Skill[] {
    return this.skills;
  }

  getName(): string {
    return this.name;
  }

  setName(name: string): void {
    this.name = name;
    console.debug(`Name changed to ${name}`);
  }

  getProfession(): Profession | null {
    return this.profession;
  }

  toString(): string {
    let str = `Name: ${this.name}`;
    if (this.profession !== null) {
      str += `, Profession: ${getProfessionName(this.profession)}`;
    }
    for (const skill of this.skills) {
      str += `, Skill: ${getSkillName(skill)}`;
    }
    str += this.isMale ? ", Gender: Male" : ", Gender: Female";
    return str;
  }

  setIsMale(isMale: boolean): void {
    this.isMale = isMale;
    console.debug(`Gender changed to ${isMale ? "Male" : "Female"}`);
  }

  getIsMale(): boolean {
    return this.isMale;
  }

  getInventory(): Inventory {
    return this.inventory;
  }

  addItemsToInventory(items: Item[]): void {
    this.inventory.addItemsToInventory(items);
  }

  addItemToInventory(item: Item): void {
    this.inventory.addItemToInventory(item);
  }

  removeItemFromInventory(itemType: ItemType, quantity: number): Item[] | null {
    return this.inventory.removeItemFromInventory(itemType, quantity);
  }

  getHealth(): Condition {
    return this.health;
  }

  decreaseHealth(amount: number): void {
    this.health.decrease(amount);
  }

  increaseHealth(amount: number): void {
    this.health.increase(amount);
  }

  getConditionPercentage(): number {
    return this.health.getPercentage();
  }

  canGetItem(itemType: ItemType, numberOf: number): boolean {
    return this.inventory.canGetItems(itemType, numberOf);
  }

  getMaxSize(): number {
    return Person.MAX_INVENTORY_SIZE;
  }

  getMaxWeight(): number {
    return Person.MAX_INVENTORY_WEIGHT;
  }

  getIsLeader(): boolean {
    return this.isLeader;
  }

  setLeader(isLeader: boolean): void {
    this.isLeader = isLeader;
  }

  getWeight(): number {
    return this.inventory.getWeight();
  }

  increaseSkillPoints(amount: number): void {
    this.skillPoints.increase(amount);
  }

  killForFood(): Item[] {
    const numberOf = Math.floor(this.weight / 5);
    const itemList: Item[] = [];
    for (let i = 0; i < Math.floor(numberOf / 2); i++) {
      itemList.push(new Item(ItemType.STRANGEMEAT));
    }
    return itemList;
  }

  getCondition(): Condition {
    return this.health;
  }

  isDead(): boolean {
    return this.dead;
  }

  setDead(dead: boolean): void {
    this.dead = dead;
  }

  makeRandom(): void {
    const professions = Object.values(Profession);
    this.setProfession(
      professions[Math.floor(Math.random() * professions.length)],
    );

    const personSkill: Skill[] = [];
    let skillPoints = 0;

    const allSkills = Object.values(Skill);
    let tempSkill = allSkills[Math.floor(Math.random() * allSkills.length)];
    while (
      tempSkill !== Skill.NONE &&
      personSkill.length < 3 &&
      skillPoints + getSkillCost(tempSkill) < 120
    ) {
      if (!personSkill.includes(tempSkill)) {
        personSkill.push(tempSkill);
        skillPoints += getSkillCost(tempSkill);
      }

      tempSkill = allSkills[Math.floor(Math.random() * allSkills.length)];
    }

    for (const skill of personSkill) {
      this.addSkill(skill);
    }

    this.inventory.addRandomItems();
  }

  getSkillsAsString(): string {
    let str = "";
    let isFirst = true;
    for (const skill of this.skills) {
      if (skill !== Skill.NONE) {
        if (!isFirst) {
          str += ", ";
        }
        str += getSkillName(skill);
        if (isFirst) {
          isFirst = false;
        }
      }
    }
    return str;
  }
}
